### DISCUSSION TITLE
Sending binary files to ESP32

### DISCUSSION THREAD

**Discussion by MichaSU99** [micropython/micropython#13678] [ESP32] [answered]
Hello @all,

I try to send binary files (aprox. 40k) to my ESP32 (ESP32-DevKit C V4 with ESP32-WROOM).
Therefor I send a script with my data to the ESP via pySerial like this one:

```
with open("{file_name}", "wb") as f:
    f.write(bytearray.fromhex("A04D32....."))
```
The Script is utf-8 encoded, except the content data, which is convertet as hex-string, and sent as 1k chunks to a raw_repl commandline.

This works for small files (approx. 20k)
Bigger files result in a failure: "MemoryError: memory allocation failed, allocating xxx bytes"

There must be a better way to do this (with thonny, I can transfer bigger files).
Does anyone know how to do?
Or is there a way to directly write into ESP's Filesystem (not SD-Card)?

Thanks in advance for help...
Michael.

PS: I'm a hobbyist, not a professional and experienced programmer

**Comment by robert-hh** [comment, marked as answer]
If you try to send the files from the PC, you can use mpremote as the standard tool. If you have a WiFi connection, you can as well use FTP (https://github.com/robert-hh/FTP-Server-for-ESP8266-ESP32-and-PYBD). Obviously, you have to get the uftpd.py file to the board first, using e.g. mpremote.

**Comment by MichaSU99** [comment]
Hi Robert,
thanks for your hints. transport_serial.py from mpremote seems to be perfect for learning and doing.
ftp-server also very interesting!
Thanks a lot!
Michael.
